### PR TITLE
Fix moreh_norm_backward for p = ±inf

### PR DESCRIPTION
**Pull request by heroxcreed**
# Fix `moreh_norm_backward` for `p = ±inf`

Addresses #56248.

## Root cause

`get_floored_p_and_decimal_and_p_is_negative` (`moreh_norm_backward_program_factory.cpp:37-44`) does `floor(p)` / `p - floor(p)` unconditionally and casts the floored magnitude to `uint32_t`. For `p = ±inf` this is `floor(±inf) = ±inf` → the cast is undefined (0 on this path), and the kernel's power ladder computes `x^(p-1)` from a garbage exponent. Result on device (Blackhole ttsim, bf16):

| p | stock `norm_backward` dx | torch autograd dx |
|---|---|---|
| `+inf` | `-inf` / garbage at the max-\|x\| element, 0 elsewhere | `sign(x)·dy` at argmax(\|x\|), 0 elsewhere |
| `-inf` | all zeros | `sign(x)·dy` at argmin(\|x\|), 0 elsewhere |

## Fix

Mirror the forward op's special-casing pattern (`moreh_norm_program_factory_{nc,h,w}_other.cpp`, `IS_ZERO` / `MINUS_INF`):

- **factory**: when `std::isinf(p)`, emit a `NORM_INF` define instead of feeding ±inf into the floored-exponent path.
- **kernel** (`#ifdef NORM_INF`): exact sub-gradient, reusing the kernel's existing bcast idioms —
  1. `tmp4 = |x|`
  2. `tmp5 = |x| - y` (y broadcast along the reduced dims; A = full tile, B = y, same operand convention as every stock call site)
  3. `tmp4 = eq(tmp5, 0)` — the arg-extremum support. One branch serves both signs: y equals `max|x|` for `+inf` and `min|x|` for `-inf`, so only the equality matters.
  4. `tmp5 = sign(x) * dy` (dy broadcast along the reduced dims)
  5. `dx = tmp5 * tmp4`

The finite-p power ladder is untouched (`#else`).

## Verification (Blackhole ttsim, bf16, torch autograd reference)

±inf matrix, all `max|err| = 0`:

| case | stock | fixed |
|---|---|---|
| 1D `dim=0`, `p=±inf` (incl. `dy=5`) | garbage / 0 | **PASS** (err 0) |
| 2D `dim=1` (W, COL bcast) `p=±inf` | garbage / 0 | **PASS** (err 0) |
| 2D `dim=0` (H, ROW bcast) `p=±inf` | garbage / 0 | **PASS** (err 0) |
| 4D `[5,8,78,77]`, `dim=0/1/2/3`, `p=±inf` | garbage / 0 | **PASS** (err ≤ 0.002) |

Zero regression on the untouched path: the existing finite-order backward suite (`p ∈ {2.0, 2.5, -2.5}`, 138 cases) passes unchanged.

New regression tests: `test_moreh_norm_backward_p_inf` — `p ∈ {inf, -inf}` × dims `{N, C, H, W}` × keepdim on `[5,8,78,77]`, with per-group-unique `|x|` (odd/256 magnitudes, exact in bf16) so torch's reference is unambiguous; keepdim=False follows the official `check_dim` skip rules (H/W skipped, same as the finite-p suite).

## Known limitation (ties)

When the arg-extremum of `|x|` is tied across T elements, torch splits `dy` evenly (`dy/T` each); this fix gives every tie member the full `dy`. Resolving ties exactly requires a cross-core reduction of the tie count — noted as future work. The regression tests construct tie-free inputs, so this divergence is explicit rather than latent.

## Repro (before)

```python
import torch, ttnn
x = torch.tensor([-1.0, 3.0, -2.5, 2.0], requires_grad=True)
y = torch.norm(x, p=float("inf"), dim=0, keepdim=True)
dy = torch.ones_like(y)
y.backward(dy)  # x.grad == [0, 1, 0, 0]

dev = ttnn.open_device(device_id=0)
tx = x.detach().bfloat16(); ty = y.detach().bfloat16(); tdy = dy.bfloat16()
nx = ttnn.from_torch(tx, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=dev)
ny = ttnn.from_torch(ty, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=dev)
ndy = ttnn.from_torch(tdy, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=dev)
dx = ttnn.operations.moreh.norm_backward(nx, ny, ndy, float("inf"), dim=0, keepdim=True)
print(ttnn.to_torch(dx))  # stock: [-inf, 0, 0, 0]  → fixed: [0, 1, 0, 0]
```

### CODE REVIEW

**Review by bbradelTT — APPROVED** [review]

**Comment by bbradelTT** [comment]
@heroxcreed could you please rebase?

**Comment by heroxcreed** [comment]
Rebased onto main (e37ab0ff1f). The patch itself is unchanged by the rebase: one commit, three files, +184 lines. Sorry for the lag on this one — I was heads-down on a batch of SFPU patches. CI is re-running now.

**Comment by heroxcreed** [comment]
Applied black 23.10.1 formatting to the new test file — the pre-commit job on the
previous attempt failed only on that (long lines were wrapped). No behaviour change;
the patch itself is untouched.
